snuba/query/parser/expressions.py

```diff
@@ -21,22 +21,26 @@
 # This root element is needed because of the ambiguity of the aggregation
 # function field which can mean a clickhouse function expression or the simple
 # name of a clickhouse function.
-root_element    = function_call / function_name
-expression      = function_call / simple_term
-parameters_list = parameter* (expression)
-parameter       = expression space* comma space*
-function_call   = function_name open_paren parameters_list? close_paren (open_paren parameters_list? close_paren)?
-simple_term     = quoted_literal / numeric_literal / column_name
-literal         = ~r"[a-zA-Z0-9_\.:-]+"
-quoted_literal  = "'" string_literal "'"
-string_literal  = ~r"[a-zA-Z0-9_\.:-]*"
-numeric_literal = ~r"-?[0-9]+(\.[0-9]+)?(e[\+\-][0-9]+)?"
-column_name     = ~r"[a-zA-Z_][a-zA-Z0-9_\.]*"
-function_name   = ~r"[a-zA-Z_][a-zA-Z0-9_]*"
-open_paren      = "("
-close_paren     = ")"
-space           = " "
-comma           = ","
+root_element          = function_call / function_name / top_arithm_expression
+expression            = low_pri_arithmetic / function_call / simple_term
+top_arithm_expression = open_paren low_pri_arithmetic close_paren
```

Comment:
For the first point - testing a string that looks like an arithmetic expression. I think this is a great idea, and I will definitely try to integrate it asap. Currently though, in this PR, the parser does not support subtraction or division. As per my knowledge, dates can be represented with the "-" or "/" signs, so perhaps once the next round of modifications to the grammar are applied, I can add this as a test case.

Second point - currently, a top-level arithmetic expression needs to be framed by a set of parentheses. This means that if the root_element is an arithmetic expression itself, it will be in parentheses (for example, (2 * 3 + 4) ). This does not mean that each unit of arithmetic ( 2 * 3 or 3 + 4 etc) are in parentheses. The top-level parentheses were added in order to avoid confusion in cases like "a + b," where "a" and "b" could be mistakenly resolved as function_name(s). So, to distinguish the fact that something like "a + b" should be recognized as an arithmetic expression, we decided that top-level arithmetic expressions need a differentiating factor - the parentheses. Hope this was a clear explanation!

Reply:
> For the first point - testing a string that looks like an arithmetic expression. I think this is a great idea, and I will definitely try to integrate it asap

I don't think this is needed. String literal are always quoted. So, in an expression `2020-06-06` would be basically invalid (or at best be an arithmetic expression). To pass a string literal that represents a date you do this `'2020-06-06'`.

Reply:
Yeah. I am talking about quoted strings. The reason I mentioned having a test for this is that I noticed arithmetic expressions takes priority over quoted literals now (line 25).

Reply:
Thanks for the explanation about the parens. I'm not sure I follow why `a + b` would be resolved as function names though since function name is not allowed inside arithmetic expression. Only function call is from what I can see, and that wouldn't match the same pattern.

Reply:
About the date it was just an example of a string that might have a arithmetic expression in it. You can test any string that contains the "+" or "*" characters.

Reply:
I think the reason why "a + b" would be resolved as function names is due to the fact that function_name is in the root_element. Hence, the parser would see "a" first and that does match with function_name grammar rule.

Reply:
> About the date it was just an example of a string that might have an arithmetic expression in it. You can test any string that contains the "+" or "*" characters in the middle.

This is another reason why the top-level parentheses are being used. "2020-06-06" will not be recognized as an arithmetic expression unless we input it like: "(2020-06-06)."

Reply:
Also, string literals were not part of the root_element in the original parser and still are not. So if we were to input "2020-06-06," it would be resolved as a function_name.

Reply:
Actually @lynnagara made me notice one thing that we can fix: `(column_name)` now would be a valid aggregation. It should not, so adding full arithmetic expressions at the top level inn parentheses has a problem.

What about this, that would simplify the grammar and make it easier to port to snql:
1) remove `function_name` entirely from the grammar at the top. Now at the top you have only arithmetic expression.
2) deal with the function name before line 152. just check if the string you are about to provide to the parser can actually be a function name. (just apply the same regex with the `re` module. `[a-zA-Z_][a-zA-Z0-9_]*`
3) if it is a valid function name, then treat it as such and go what happens at line 206.
4) if not, then apply the parser as you do today. Now every arithmetic expression including simple terms will be valid.

What do you think ?

Reply:
Sounds good. I simplified the grammar and some of the visitor code. Let me know what you think!

Reply:
I think that looks a lot cleaner, thanks